cloudmarker/clouds/gcpcloud.py

```diff
@@ -226,3 +192,112 @@ def done(self):
         up tasks associated with the :class:`GCPCloud` plugin. This
         may change in future.
         """
+
+
+def _get_normalized_firewall_rules(firewall):
+    """Split a firewall record into multiple firewall rules.
+
+    A firewall record in GCP has a top-level key named either
+    ``allowed`` or ``denied``. The value of this key is a list of
+    allowed or denied rules. Each such rule contains the name of allowed
+    or denied protocol along with a list of allowed or denied port
+    ranges.
+
+    In order to make it easier to write event plugins to detect security
+    issues in firewall, we generate a new firewall rule record for each
+    allowed or denied rule we find in the value of ``allowed`` or
+    ``denied`` keys in a firewall record.
+
+    Arguments:
+        firewall (dict): Raw firewall record retrieved from GCP.
+    """
+    allow_rules = firewall.get('allowed')
+    if allow_rules is not None:
+        for rule in allow_rules:
+            firewall_rule = _get_normalized_firewall_rule(firewall, rule)
+            firewall_rule['com']['access'] = 'allow'
+            yield firewall_rule
+
+    deny_rules = firewall.get('denied')
+    if deny_rules is not None:
+        for rule in deny_rules:
+            firewall_rule = _get_normalized_firewall_rule(firewall, rule)
+            firewall_rule['com']['access'] = 'allow'
+            yield firewall_rule
+
+
+def _get_normalized_firewall_rule(firewall, rule):
+    """Create a normalized firewall rule record.
+
+    Arguments:
+        firewall (dict): Raw firewall record retrieved from GCP.
+        rule (dict): Raw allowed or denied rule in ``firewall``.
+
+    Returns:
+        dict: A normalized firewall rule record with ``com`` bucket
+            populated with firewall rule properties in common notation.
+
+    """
+    record = {
+        'raw': rule,
+        'ext': {
+            'cloud_type': 'gcp',
+            'record_type': 'firewall_rule',
+            'firewall_id': firewall.get('id'),
+            'firewall_link': firewall.get('selfLink'),
+        },
+        'com': {
+            'cloud_type': 'gcp',
+            'record_type': 'firewall_rule',
+            'reference': firewall.get('selfLink'),
+
+            'enabled': not firewall.get('disabled'),
+
+            'direction':
+                _get_normalized_firewall_direction(firewall),
+
+            'protocol':
+                _get_normalized_firewall_protocol(rule),
+
+            'source_addresses': firewall.get('sourceRanges'),
+
+            # If the 'ports' key is missing in an allowed/denied rule,
+            # it means all ports are allowed/denied.
+            'destination_ports': rule.get('ports', ['0-65535'])
+        }
+    }
+    return record
+
+
+def _get_normalized_firewall_direction(firewall):
+    rule_name = firewall.get('name')
+    direction = firewall.get('direction')
+
+    if direction is None:
```

Comment:
Will there be any case when the direction of firewall rule in either of Azure or GCP cloud will be `None`?

Reply:
There should not be any case where `direction` is `None`. However, the data returned to us by Azure is not in our control. When we are dealing with data that is not in our control, it's a good practice to write our code as defensively as possible with as little assumptions as possible.

In the above code, we do not make any assumptions about whether `direction` key is present or not in the response from GCP Cloud. With only about three lines of overhead, we are able to gracefully handle a scenario of missing `direction` key. So I think these three lines of overhead are worth it. It avoids the possibility of the service crashing unexpectedly if some day for some reason the `direction` key goes missing in the GCP response.